text/0000-api-evolution.md

```diff
@@ -0,0 +1,748 @@
+traits are permitted to add new type parameters as long as defaults are provided
+(which is backwards compatible).
+
+### Trait implementations
+
+#### Major change: implementing any "fundamental" trait.
+
+A [recent RFC](https://github.com/rust-lang/rfcs/pull/1023) introduced the idea
+of "fundamental" traits which are so basic that *not* implementing such a trait
+right off the bat is considered a promise that you will *never* implement the
+trait. The `Sized` and `Fn` traits are examples.
+
+The coherence rules take advantage of fundamental traits in such a way that
+*adding a new implementation of a fundamental trait to an existing type can
+cause downstream breakage*. Thus, such impls are considered major changes.
+
+#### Minor change: implementing any non-fundamental trait.
+
+Unfortunately, implementing any existing trait can cause breakage:
+
+```rust
+// Crate A
+    pub trait Trait1 {
+        fn foo(&self);
+    }
+
+    pub struct Foo; // does not implement Trait1
+
+// Crate B
+    use crateA::Trait1;
+
+    trait Trait2 {
+        fn foo(&self);
+    }
+
+    impl Trait2 for crateA::Foo { .. }
+
+    fn use_foo(f: &crateA::Foo) {
+        f.foo()
+    }
+```
+
+If crate A adds an implementation of `Trait1` for `Foo`, the call to `f.foo()`
+in crate B will yield a dispatch ambiguity (much like the one we saw for
+defaulted items). Thus *technically implementing any existing trait is a
+breaking change!* Completely prohibiting such a change is clearly a non-starter.
+
+However, as before, this kind of breakage is considered "minor" by the
+principles of this RFC (see "Adding a defaulted item" above).
+
+### Inherent implementations
+
+#### Minor change: adding any inherent items.
+
+Adding an inherent item cannot lead to dispatch ambiguity, because inherent
+items trump any trait items with the same name.
+
+However, introducing an inherent item *can* lead to breakage if the signature of
+the item does not match that of an in scope, implemented trait:
+
+```rust
+// Crate A
+    pub struct Foo;
+
+// Crate B
+    trait Trait {
+        fn foo(&self);
+    }
+
+    impl Trait for crateA::Foo { .. }
+
+    fn use_foo(f: &crateA::Foo) {
+        f.foo()
+   }
+```
+
+If crate A adds a method:
+
+```rust
+impl Foo {
+    fn foo(&self, x: u8) { ... }
+}
+```
+
+then crate B would no longer compile, since dispatch would prefer the inherent
+impl, which has the wrong type.
+
+Once more, this is considered a minor change, since UFCS can disambiguate (see
+"Adding a defaulted item" above).
+
+It's worth noting, however, that if the signatures *did* happen to match then
+the change would no longer cause a compilation error, but might silently change
+runtime behavior. The case where the same method for the same type has
+meaningfully different behavior is considered unlikely enough that the RFC is
+willing to permit it to be labeled as a minor change -- and otherwise, inherent
+methods could never be added after the fact.
```

Comment:
hmm ... I wonder if this could provide motivation for some way to `use` a type without bringing its inherent impls into scope.